crates/svm-cli/tests/spawn_app_tests.rs

```diff
@@ -0,0 +1,216 @@
+use regex::Regex;
+
+use svm_cli::{cli, wasm_value};
+use svm_common::Address;
+
+struct SpawnAppTestCase {
+    version: String,
+    template_addr: String,
+    ctor_index: String,
+    ctor_buf: String,
+    ctor_args: Vec<String>,
+}
+
+#[test]
+fn encode_decode() {
+    let cases = vec![
+        SpawnAppTestCase {
+            version: String::from("0"),
+            template_addr: String::from("00aa00aa00aa00aa00aa00aa00aa00aa00aa00aa"),
+            ctor_index: String::from("0"),
+            ctor_buf: String::from("11bb11bb12345678"),
+            ctor_args: vec![String::from("10i32"), String::from("20i64")],
+        },
+        SpawnAppTestCase {
+            version: String::from("0"),
+            template_addr: String::from("00aa00aa00aa00aa00aa00aa00aa00aa00aa00aa"),
+            ctor_index: String::from("12"),
+            ctor_buf: String::from(""),
+            ctor_args: vec![String::from("1073741824i64"), String::from("0i32")],
+        },
+    ];
+
+    for case in cases {
+        test_encode_decode(case)
+    }
+}
+
+fn test_encode_decode(case: SpawnAppTestCase) {
+    let tempfile_path = tempfile::NamedTempFile::new().unwrap();
+    let tempfile_path = tempfile_path.path().to_str().unwrap();
+
+    let output_path = tempfile_path;
+    let mut input = vec![
+        "myprog",
+        "encode",
+        "spawn_app",
+        output_path,
+        &case.version,
+        &case.template_addr,
+        &case.ctor_index,
+        &case.ctor_buf,
+    ];
+    for i in 0..case.ctor_args.len() {
+        input.push(&case.ctor_args[i]);
+    }
+    let matches = cli::new_app().get_matches_from(input);
+    let output = cli::process(matches).unwrap();
+
+    let re = Regex::new(r"Wrote (\d+) bytes to (.*)").unwrap();
+    let cap = re.captures(&output).unwrap();
+    assert_eq!(&cap[2], output_path);
+
+    let data_path = output_path;
+    let input = vec!["myprog", "decode", "spawn_app", data_path];
+    let matches = cli::new_app().get_matches_from(input);
+    let output = cli::process(matches).unwrap();
+
+    let re = Regex::new(
+        r"Version: (.*)\nTemplate: (.*)\nctor_idx: (\d+)\nctor_buf: (.*)\nctor_args: (.*)",
+    )
+    .unwrap();
+    let caps = re.captures(&output).unwrap();
+
+    assert_eq!(&caps[1], case.version);
+    assert_eq!(&caps[2], {
+        let template_addr = hex::decode(case.template_addr).unwrap();
```

Comment:
better to introduce some helper function (maybe a local function within the test).

```rust
let template_addr = normalize(..);
assert_eq!(&caps[2], template_addr);
```